android/app/src/main/java/io/peggyjo/peggymeter/PDKListener.java

```diff
@@ -2,17 +2,78 @@
 
 import android.content.Context;
 import android.os.Bundle;
+import android.support.annotation.NonNull;
 import android.util.Log;
-
 import com.audacious_software.passive_data_kit.generators.Generators;
 
-public class PDKListener implements Generators.GeneratorUpdatedListener {
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+
+
+public class PDKListener implements Generators.GeneratorUpdatedListener,
+        OnCompleteListener<Void> {
+    private final HashMap<String, Long> timestamps = new HashMap<>();
+    private final FirebaseFirestore db = FirebaseFirestore.getInstance();
+    private final String uid = FirebaseAuth.getInstance().getCurrentUser().getUid();
+
     public void onGeneratorUpdated(String identifier, long timestamp, Bundle data) {
-        Log.i("PDK listener", "id: " + identifier + " | ts: " + timestamp + " | data.size = " + data.size());
-        // TODO: implement uploading to Firebase.
+        boolean shouldUpload = shouldUpload(identifier, timestamp);
+
+        if (shouldUpload) {
+            String docKey = uid + "-" + identifier + "-" + timestamp;
+            uploadData(data, docKey);
+        }
     }
+
     public void initialize(Context context) {
         Context appCtx = context.getApplicationContext();
         Generators.getInstance(appCtx).addNewGeneratorUpdatedListener(this);
     }
+
+    @Override
+    public void onComplete(@NonNull Task<Void> task) {
+        if (task.isSuccessful()) {
+            Log.i("PDKListener", "File write succeeded.");
+        } else {
+            Log.i("PDKListener", "Error writing file.");
+        }
+    }
+
+    private long calculateDeltaMins(Date current, Date previous) {
+        long diffMills = current.getTime() - previous.getTime();
+        long deltaMins = TimeUnit.MINUTES.convert(diffMills, TimeUnit.MILLISECONDS);
+        return deltaMins;
+    }
+
+    private boolean shouldUpload(String identifier, long timestamp) {
+        if (!timestamps.containsKey(identifier)) {
+            timestamps.put(identifier, timestamp);
+            return true;
+        } else {
```

Comment:
You don't need else here because the function returns in the if block.
Generally, the less indentation the better.
E.g. you can list all cases for returning true and at the end return false.

if (cond1) {
  return true;
}
...
if (cond2) {
  return true;
}
return false;